### DISCUSSION TITLE
Dynamically setting mods to match difficulty preferences

### DISCUSSION THREAD

**Discussion by NafeeJ** [ppy/osu#23262] [Ideas]
It would be a nice feature to just set preferences to use the mods to adjust all maps to them.

For example, I would choose SR 6, AR 9, 200 BPM and it would just use DT or Diff Adjust to automatically make it like that for all maps I select

**Comment by monochrome22** [comment]
The Difficulty Adjust mod already does this for AR OD CS HP. If you set it as AR 9, all maps will be AR 9. Except if you use DT or HT, in which case the AR is affected by speed adjust. I agree it should force the AR of DA instead, but this seems difficult, see https://github.com/ppy/osu/discussions/14170 and https://github.com/ppy/osu/discussions/15249#discussioncomment-2399529. Also, peppy seems to be against it: https://github.com/ppy/osu/discussions/15249#discussioncomment-2399914. 
One option would be to make it automatically adjust the AR and OD to compensate for the speed change. Meaning, if the player sets AR 10 with DA and 1.5x speed with DT, then DA would automatically change AR to 8.5. Could be a toggle in DA, "Automatically adjust settings to speed changes".  For this to work UX wise, first, speed adjust mods should display the real AR and OD, as proposed in https://github.com/ppy/osu/issues/10580. Then, if the toggle is enabled, the settings sliders would be limited to what ARs and ODs can be set. For example, if you enable 0.5x HT, then the AR and OD would only go up to 9. There should be an explanation when hovering the toggle: "The available values are limited by Double Time and Half Time settings".  

For BPM, see https://github.com/ppy/osu/discussions/22616.
Having DT/HT adjusted to match a certain SR would have the same problems as setting a manual BPM. Manual BPM will probably get added at some point though, since it would be a good feature and it's decently requested, so maybe adding SR matching could be considered as well. 

If they add adjusting DA to DT/HT changes and setting manual BPM, then you can use those and star=6 to find all the maps that fit your preferences. This is assuming that it's possible to know what the map's star rating will be at different speeds without clicking on it. 
https://github.com/ppy/osu/issues/19947 would also need to be added.

**Reply by Rekunan** [reply]
Adding on, using the example, in order to have DA automatically adjust the map to be 6 stars, it would have to change the circle size. However, what if there isn't a set bpm; how should osu! determine what to automatically adjust to fit within the star rating?

**Reply by monochrome22** [reply]
> Adding on, using the example, in order to have DA automatically adjust the map to be 6 stars, it would have to change the circle size. 

Correct. I didn't talk about DA adjusting to match a star rating because I thought it would be a weird feature. It feels more correct to only have DT/HT matching a star rating instead. Also, having both DA and DT/HT automatically adjusting to match a star rating would be complicated UX wise. 


> However, what if there isn't a set bpm; how should osu! determine what to automatically adjust to fit within the star rating?

The solution to this is written in the BPM discussion I linked: "Additionally, for variable BPM maps it would have to use the most common BPM (i.e. 'mostly X bpm' in the beatmap details)."

**Reply by Rekunan** [reply]
When I mean "set bpm" I mean as in a set rate, because yeah it'd be a complicated UX issue to be able to have DA and rate changes auto adjust, compared to just one or the other.

**Comment by exformation** [comment]
Thought of something similar and I think it would be a very useful feature. Here's my initial message about it from the discord: 

Would it be possible to make a mod that sets all maps to a certain BPM? Ex: I set it to 220 and every map automatically speeds up or slows down to match 220 for its main BPM. It:
- Might want to filter out any maps that would require more than a 2.0x or less than a 0.5x speed multiplier.
- Might want to have a toggle for whether or not people want slowed down maps to show up.
- Might need an AR setting, so that all the maps have the same AR regardless of them being sped up or slowed down.

Doing this through DA makes a lot of sense, since you're already setting AR and what not. A BPM slider could be added to DA for this, which might be simpler than dealing with an input field. You'd just have to decide on an arbitrary range, like 150-300.
